File: c/stack.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#define MAX_STACK 6

struct STACK {
	int top;
	char data[10][MAX_STACK];
};

struct STACK tumpukanBuku;

void inisialisasi(){
	tumpukanBuku.top = -1;
}

int IsFull(){
	if(tumpukanBuku.top == MAX_STACK-1) 
		return 1; 
	else 
		return 0;
}

int IsEmpty(){
	if(tumpukanBuku.top == -1) 
		return 1; 
	else 
		return 0;
}

Push(){
	if(IsFull() != 1){
		printf("Buku yang akan ditumpuk dalam tumpukan : ");
		tumpukanBuku.top++;
		scanf("%s",tumpukanBuku.data[tumpukanBuku.top]);
	} else printf("\nTumpukan buku sudah penuh!\n");
}

Pop(){
	if(IsEmpty() != 1) {
		printf("Buku yang terambil = %s\n",tumpukanBuku.data[tumpukanBuku.top]);
		tumpukanBuku.top--;
	}
	else printf("\nTidak ada buku yang terambil, tumpukan masih kosong!\n");
}

Clear(){
	tumpukanBuku.top=-1;
	printf("\nTumpukan buku sudah kosong!\n");
}

Cetak(){
	if(IsEmpty() != 1) {
		int i;
		printf("Tumpukan Buku : \n");
		for(i=tumpukanBuku.top;i>=0;i--){
			printf("%s\n",tumpukanBuku.data[i]);
		}
	} else printf("\nTumpukan buku masih kosong!\n");
}

int main(){
	int pilihan;
	inisialisasi();
	char dt[10];
	do{
		printf("1. push\n");
		printf("2. pop\n");
		printf("3. cetak\n");
		printf("4. clear\n");
		printf("5. exit\n");
		printf("Pilihan : ");scanf("%d",&pilihan);
		
		switch(pilihan){
			case 1: 
				Push();
				break;
		
			case 2: 
				Pop();
				break;

			case 3: 
				Cetak();
				break;

			case 4: 
				Clear();
				break;
			case 5:
				return 0;		
			default:
				printf("Tidak ada dalam pilihan\n");
		}
	} while(pilihan>=1 && pilihan<=5);
	getchar();
}

File: pascal/Stack.pas
PROGRAM Stack;

TYPE 
	Stack = RECORD
		top : integer;
		data : ARRAY [1..6] of string[30];
	END;

VAR 
	TumpukanBuku : Stack;
	key : integer;

PROCEDURE InisialisasiStack;
BEGIN
	TumpukanBuku.top := 0;
END;

FUNCTION IsEmpty() : boolean;
BEGIN
	IF TumpukanBuku.top = 0 THEN return TRUE
	ELSE return FALSE;
END;

FUNCTION IsFull() : boolean;
BEGIN
	IF TumpukanBuku.top = 6 THEN return TRUE
	ELSE return FALSE; 
END;

PROCEDURE Clear;
BEGIN
	TumpukanBuku.top := 0;
END;

PROCEDURE Push;
BEGIN
	VAR elemen : string[30];
	write('Elemen yang akan ditumpuk : ');
	readln(elemen);
	TumpukanBuku.top := TumpukanBuku.top + 1;
	TumpukanBuku.data[TumpukanBuku.top] := elemen;
	writeln('Memasukkan elemen : ',elemen,' pada puncak tumpukan');
END;

PROCEDURE Pop;
BEGIN
	writeln('Mengambil elemen : ',TumpukanBuku.data[TumpukanBuku.top],' dari puncak tumpukan');
	TumpukanBuku.top := TumpukanBuku.top - 1;
END;

PROCEDURE Cetak;
BEGIN
	VAR i : integer;
	writeln('Tumpukan Buku : ');
	FOR i:=TumpukanBuku.top DOWNTO 1 DO
	BEGIN
		write(TumpukanBuku.data[i],' ');
	END;
	writeln;
END;

{program utama}
BEGIN
	InisialisasiStack;

		writeln('1. Push');
		writeln('2. Pop');
		writeln('3. Print');
		writeln('4. Clear');
		writeln('5. Exit');
		write('Masukkan pilihan anda : ');
	read(key);
		

			CASE key of
				1 : Push;
				2 : Pop;
				3 : Cetak;
				4 : Clear;
				else writeln('Masukkan pilihan 1 - 5');
			END;
		

END.	
